src/test/lib/concurrency/stress_test.cpp

```diff
@@ -263,6 +276,178 @@ TEST_F(StressTest, NodeSchedulerStressTest) {
+      if (!queue) {
+        continue;
+      }
+      EXPECT_EQ(queue->semaphore.availableApprox(), 0);
+    }
+
+    Hyrise::get().scheduler()->schedule_tasks(waiting_jobs);
+    // Wait a bit for workers to pull jobs and decrement semaphore.
+    std::this_thread::sleep_for(5 * CORES_PER_NODE * SLEEP_TIME);
+
+    for (const auto& queue : node_queue_scheduler->queues()) {
+      if (!queue) {
+        continue;
+      }
+      EXPECT_EQ(queue->semaphore.availableApprox(), job_count - CORES_PER_NODE);
+    }
+
+    wait_flag = false;
+    Hyrise::get().scheduler()->wait_for_tasks(waiting_jobs);
+
+    for (const auto& queue : node_queue_scheduler->queues()) {
+      if (!queue) {
+        continue;
+      }
+      EXPECT_EQ(queue->semaphore.availableApprox(), 0);
+    }
+    EXPECT_EQ(job_count, counter);
+  }
+}
+
+// Similar to test above, but here we make tasks dependent of each other which means only non-dependent tasks will be
+// scheduled.
+TEST_F(StressTest, NodeQueueSchedulerSemaphoreIncrementsDependentTasks) {
+  constexpr auto DEPENDENT_JOB_TASKS_LENGTH = size_t{10};
+  constexpr auto SLEEP_TIME = std::chrono::milliseconds{1};
+  const auto job_count = CORES_PER_NODE * 32;
+
+  for (const auto& fake_numa_topology : FAKE_SINGLE_NODE_NUMA_TOPOLOGIES) {
+    Hyrise::get().topology.use_fake_numa_topology(fake_numa_topology);
+    auto node_queue_scheduler = std::make_shared<NodeQueueScheduler>();
+    Hyrise::get().set_scheduler(node_queue_scheduler);
+
+    auto counter = std::atomic<uint32_t>{0};
+    auto wait_flag = std::atomic<bool>{true};
+
+    auto waiting_jobs = std::vector<std::shared_ptr<AbstractTask>>{};
+    waiting_jobs.reserve(job_count);
+    for (auto job_id = size_t{0}; job_id < job_count; ++job_id) {
+      waiting_jobs.emplace_back(std::make_shared<JobTask>([&] {
+        while (wait_flag) {
+          std::this_thread::sleep_for(SLEEP_TIME);
+        }
+        ++counter;
+      }));
+      // We create runs of dependent jobs and set the current job as a predecessor of the previous job.
+      if (job_id % DEPENDENT_JOB_TASKS_LENGTH != 0) {
+        waiting_jobs.back()->set_as_predecessor_of(waiting_jobs[job_id - 1]);
+      }
+    }
+
+    for (const auto& queue : node_queue_scheduler->queues()) {
+      if (!queue) {
+        continue;
+      }
+      EXPECT_EQ(queue->semaphore.availableApprox(), 0);
+    }
+
+    Hyrise::get().scheduler()->schedule_tasks(waiting_jobs);
+    // Wait a bit for workers to pull jobs and decrement semaphore.
+    std::this_thread::sleep_for(5 * CORES_PER_NODE * SLEEP_TIME);
+
+    for (const auto& queue : node_queue_scheduler->queues()) {
+      if (!queue) {
+        continue;
+      }
+      EXPECT_EQ(queue->semaphore.availableApprox(),
+                static_cast<size_t>(
+                    std::ceil(static_cast<float>(job_count) / static_cast<float>(DEPENDENT_JOB_TASKS_LENGTH))) -
+                    CORES_PER_NODE);
```

Comment:
Tremendous to put this in one statement. Maybe pull into variable and elaborate a bit?

Reply:
What about this?

```
      // The number of scheduled jobs depends on DEPENDENT_JOB_TASKS_LENGTH (see job definition above; due the jobs
    // dependencies, jobs are only scheduled when they have no predecessors).
    const auto executable_jobs = static_cast<float>(job_count) / static_cast<float>(DEPENDENT_JOB_TASKS_LENGTH);
    for (const auto& queue : node_queue_scheduler->queues()) {
      if (!queue) {
        continue;
      }

      // We started scheduled jobs, which block all workers due to `wait_flag` being true. Thus, the semaphore should be
      // reduced by the number of workers (i.e., CORES_PER_NODE).
      EXPECT_EQ(queue->semaphore.availableApprox(), static_cast<size_t>(std::ceil(executable_jobs)) - CORES_PER_NODE);
    }
```

Reply:
Like that more. Should it be `due _to_` though?